### PR TITLE
Add Maven Wrapper to simplify installation for deployment

### PR DESCRIPTION
**Pull request by markmatney**
The RHEL 7 server we need to deploy to does not have Maven installed.
This change gets around that with mvnw, taking the same approach as
av-pairtree. The added files were generated with the command

  mvn -N io.takari:maven:wrapper

using Maven 3.6.0.

### CODE REVIEW

**Review by ksclarke — DISMISSED** [review]
👍🏼  We might want to start adding that to all projects

**Comment by ksclarke** [comment]
Override/Suppress that location that Codacy is complaining about in the Checkstyle rules. We don't control that code.

**Review by DRickard — COMMENTED** [review]
Nitpick: the README still says Maven is required to build, doesn't say how to build using `mvnw`.

**Comment by markmatney** [comment]
@ksclarke: should that be a change to the parent project (to ignore these files)? Or should I edit the checkstyle.xml in this repo?
@DRickard: thanks, will fix.

**Comment by ksclarke** [comment]
> @ksclarke: should that be a change to the parent project (to ignore these files)? Or should I edit the checkstyle.xml in this repo?

You can supply a different suppressions file in the POM (that will override the upstream default). Then extend the upstream default's file to add the stuff in this particular project we want to ignore (creating a copy). Cf https://github.com/UCLALibrary/bucketeer/blob/main/pom.xml#L554 for an example (since we do this in bucketeer).

For bucketeer's copy/overrides https://github.com/UCLALibrary/bucketeer/tree/main/src/main/tools/checkstyle

**Comment by markmatney** [comment]
This feels to me like something that should be a common configuration across all projects

**Comment by ksclarke** [comment]
> This feels to me like something that should be a common configuration across all projects

It does make it easier on systems without Maven already installed, for sure.

**Comment by markmatney** [comment]
Oh, I mean the suppression of Checkstyle on `mvnw`, `mvnw.cmd`, and `.mvn/`.

**Comment by ksclarke** [comment]
> Oh, I mean the suppression of Checkstyle on `mvnw`, `mvnw.cmd`, and `.mvn/`.

Those are directories/files that get added to that suppressions file override. I was just using bucketeer as an example of how to do it. Suppressing those files in that config means Checkstyle will ignore them.

*Edit:* I get what you were saying now -- too many threads :-)

**Comment by markmatney** [comment]
Wouldn't we want to configure the parent project to suppress warnings on those files? If we're going to add mvnw to multiple projects, I wonder if it makes more sense to make the change there instead

**Comment by ksclarke** [comment]
> Wouldn't we want to configure the parent project to suppress warnings on those files? If we're going to add mvnw to multiple projects, I wonder if it makes more sense to make the change there instead

I've had mixed feelings about mvnw. It installs additional Mavens if the system already has them and I've made the assumption that developers usually have Maven installed (and don't want extra copies around). This is an odd situation where we're checking the repo out on a system vs. just installing an executable Jar from Maven central. So, I've not used it much or added it upstream before.

I was just thinking doing it locally would be faster, but if you want to do a pull request to the upstream I can merge it and do a new release of the freelib-parent. It won't hurt to have it upstream.

Either way.

*Edit:* Oh, actually... scratch that... the upstream has new PMD rules which would probably require a lot of cleanup on metagetter. We can't use the new release. Let's just do locally for now. I mean I guess I could do a new release as a bug fix off of the previous tag, but that's getting to be more work. I'd just say let's do locally for now.

**Comment by markmatney** [comment]
ok, sounds good :+1:

**Comment by markmatney** [comment]
hmm... I've tried several iterations of checkstyle suppression, to no avail. I am obviously doing something wrong, but am not sure what that is. Anyone have any insight?

One thing to note is that a local `mvn clean verify` passes on b7b7fd4 (before I attempted any of the checkstyle stuff). Is this correct behavior? Shouldn't that have been failing with the addition of the Maven Wrapper files?

File: src/main/tools/checkstyle/checkstyle-suppressions-override.xml [hunk 1]
```diff
@@ -5,6 +5,6 @@
 
 <suppressions>
   <suppress files="mvnw" checks="."/>
-  <suppress files="mvnw.cmd" checks="."/>
-  <suppress files=".mvn/wrapper/MavenWrapperDownloader.java" checks="."/>
+  <suppress files="mvnw\.cmd" checks="."/>
+  <suppress files="\.mvn[\\/]" checks="."/>
```
Comment (ksclarke) [inline src/main/tools/checkstyle/checkstyle-suppressions-override.xml:9]: Did you try without escaping the period?

File: pom.xml [hunk 2]
```diff
@@ -194,17 +194,12 @@
           <logViolationsToConsole>true</logViolationsToConsole>
```
Comment (ksclarke) [inline pom.xml]: I don't think this is related, but I'm not sure the custom overrides are required anymore. We could remove them, leaving just the suppressionsLocation

File: src/main/tools/checkstyle/checkstyle-suppressions-override.xml (hunk 1 above)
Comment (markmatney) [reply src/main/tools/checkstyle/checkstyle-suppressions-override.xml:9]: Just tried without escaping the periods, but it failed 😦

**Comment by ksclarke** [comment]
It looks like this is an older version of the checkstyle config file that's specifying its own suppressions file so maybe the setting in the pom is being ignored: https://github.com/UCLALibrary/services-metagetter/blob/a945b6a8938e5c98a6fea1fcff64606ebe9e3a4c/src/main/tools/checkstyle/checkstyle.xml#L89

You can update the parent pom version in this project to get the newer checkstyle rules.

**Comment by markmatney** [comment]
Oh wow, I totally missed that. Thank you!

**Comment by markmatney** [comment]
Ready for review again.

I have an open support ticket with Codacy in order to resolve the issue with Checkstyle.

**Review by ksclarke — APPROVED** [review]
Would you create a ticket in our system too for the issue? Might as well move it to "In Progress" too since we're waiting on assistance from Codacy.

**Comment by markmatney** [comment]
https://jira.library.ucla.edu/browse/SERV-139

**Review by DRickard — APPROVED** [review]

**Pull request merged**